src/main/java/com/aws/greengrass/tes/HttpServerImpl.java

```diff
@@ -26,30 +57,104 @@ public class HttpServerImpl implements Server {
 
     /**
      * Constructor.
-     * @param port Http server port
+     * @param port HTTP server port
      * @param credentialRequestHandler request handler for server requests
      * @param executorService executor service instance
      * @throws IOException When server creation fails
      */
     HttpServerImpl(int port, HttpHandler credentialRequestHandler, ExecutorService executorService) throws IOException {
-        httpImpl = HttpServer.create(new InetSocketAddress("localhost", port), 0);
         this.credentialRequestHandler = credentialRequestHandler;
         this.executorService = executorService;
+        this.configuredPort = port;
+        this.servers = createServers(configuredPort);
+        this.resolvedPort = this.servers.get(0).getAddress().getPort();
+    }
+
+    /**
+     * Create ipv4 and ipv6 servers, both using the specified port.
+     *
+     * @param port chosen port for all created servers
+     * @return list of http servers
+     * @throws IOException if an I/O error occurs during server creation
+     * @throws BindException if the chosen port is unavailable
+     */
+    private List<HttpServer> createServers(int port) throws IOException {
+        HttpServer serverIPv4 = null;
+        InetSocketAddress addrIPv4 = new InetSocketAddress(ADDR_IPV4, port);
+        try {
+            serverIPv4 = HttpServer.create(addrIPv4, 0);
+        } catch (SocketException e) {
+            if (canBindToAddress(ADDR_IPV4)) {
+                throw e;
+            }
+            logger.atDebug().cause(e).kv("address", addrIPv4)
+                    .log("Unable to bind HTTP server. "
+                            + "Proceeding anyway as IPv4 is likely unavailable on the host system");
+        }
+
+        HttpServer serverIPv6 = null;
+        InetSocketAddress addrIPv6 =
+                new InetSocketAddress(ADDR_IPV6,
+                // use the same port from IPv4 server
+                serverIPv4 == null ? port : serverIPv4.getAddress().getPort());
+        try {
+            serverIPv6 = HttpServer.create(addrIPv6, 0);
+        } catch (SocketException e) {
+            if (canBindToAddress(ADDR_IPV6)) {
+                if (serverIPv4 != null) { // cleanup
+                    stop(serverIPv4);
+                }
```

Comment:
can we simplify this cleanup logic by just calling `stop()`? and letting stop figure out which server(s) need to stop

Reply:
yep will rework